### PR TITLE
interactive ui for online players & matches on GET /c.ppy.sh page

### PR DESCRIPTION
**Pull request by arily**
# Describe your changes
display more information in the "bancho" page, key to judge if it's safe to restart server without login using osu client.

image
image
image

## Checklist
- [ ] The changes pass pre-commit checks (`make lint`)
- [ ] The changes follow coding style

### CODE REVIEW

**Comment by cmyui** [comment]
i knew this day would come

frontend improvements to bancho.py LOL

File: app/api/domains/cho.py [hunk 1]
```diff
@@ -84,20 +84,100 @@
 @router.get("/")
 async def bancho_http_handler():
     """Handle a request from a web browser."""
+    new_line = "\n"
+    matches = list(
+        filter(lambda match: isinstance(match, Match), app.state.sessions.matches),
+    )
```
Comment (cmyui) [inline app/api/domains/cho.py:90]: are they ever not of instance `Match`?

File: app/api/domains/cho.py [hunk 2]
```diff
@@ -84,20 +84,100 @@
...
+    players = list(filter(lambda p: not p.bot_client, app.state.sessions.players))
```
Comment (cmyui) [inline app/api/domains/cho.py:91]:
```suggestion
    players = [p for p in app.state.sessions.players if not p.bot_client]
```

Comment (cmyui) [reply app/api/domains/cho.py:91]: generally more concise

File: app/api/domains/cho.py (hunk 1 above)
Comment (arily) [reply app/api/domains/cho.py:90]: [there's None](https://github.com/arily/bancho.py/blob/9078c4f68e22d6b8f096baeef9d918d48803e2e4/app/objects/collections.py#L175)

File: app/api/domains/cho.py [hunk 3]
```diff
@@ -84,20 +84,100 @@
...
+    )
+
+
+@router.get("/online")
+async def bancho_list_user():
+    """see who's online"""
+    new_line = "\n"
+    players = list(filter(lambda p: not p.bot_client, app.state.sessions.players))
+    bots = list(filter(lambda p: p.bot_client, app.state.sessions.players))
+    id_max_length = len(str(max(p.id for p in app.state.sessions.players)))
```
Comment (cmyui) [inline app/api/domains/cho.py:117]: same can be done for these

File: app/api/domains/cho.py (hunk 1 above)
Comment (cmyui) [reply app/api/domains/cho.py:90]: oh, probably better to check for that

Comment (cmyui) [reply app/api/domains/cho.py:90]:
```py
matches = [m for m in app.state.sessions.matches if m is not None]
```

File: app/api/domains/cho.py (hunk 3 above)
Comment (cmyui) [reply app/api/domains/cho.py:117]: and the others later on

File: app/api/domains/cho.py (hunk 1 above)
Comment (arily) [reply app/api/domains/cho.py:90]: f-strings below will throw error if None is not filtered out, tested locally

File: app/api/domains/cho.py [hunk 4]
```diff
@@ -84,20 +84,100 @@
...
-                "<br>".join([f"{packet.name} ({packet.value})" for packet in packets]),
-            ),
-        ).encode(),
+        f"""
+<!DOCTYPE html>
+<body style="font-family: monospace;  white-space: pre-wrap;"><a href="/">back</a>
+users:
+{new_line.join(
+    map(
+        lambda p: f"({str(p.id).rjust(id_max_length)}): {p.safe_name}",
```
Comment (cmyui) [inline app/api/domains/cho.py:126]: if i remember correctly, i think this should work as well for the `f"{str(x).rjust(y)}"` cases, but i don't particularly care either way
```suggestion
        lambda p: f"({p.id:>{id_max_length}}): {p.safe_name}",
```

File: app/api/domains/cho.py (hunk 1 above)
Comment (cmyui) [reply app/api/domains/cho.py:90]: yea you 100% need to remove the nones, the `if m is not None` does the filtering in my example

File: app/api/domains/cho.py (hunk 4 above)
Comment (arily) [reply app/api/domains/cho.py:128]: LGTM

Comment (arily) [reply app/api/domains/cho.py]: how to do a ljust? <: ?

Comment (cmyui) [reply app/api/domains/cho.py]: (there are many more cases of this though - my only request is to pick one style and use it consistently)

Comment (cmyui) [reply app/api/domains/cho.py]: almost - `:<`

File: app/api/domains/cho.py (hunk 3 above)
Comment (arily) [reply app/api/domains/cho.py]: fixed

File: app/api/domains/cho.py (hunk 4 above)
Comment (cmyui) [reply app/api/domains/cho.py]: in python `:` is the delimiter between the data and the format spec

e.g.
```py
balance = 123.4567
print(f"Balance: ${123:.2f}") # here, the format spec is `.2f`, for "2 decimal float"
> 123.45
```

Comment (cmyui) [reply app/api/domains/cho.py]: more format specs here https://www.w3schools.com/python/ref_string_format.asp

File: app/api/domains/cho.py (hunk 1 above)
Comment (arily) [reply app/api/domains/cho.py]: updated

File: app/api/domains/cho.py (hunk 4 above)
Comment (arily) [reply app/api/domains/cho.py]: updated

**Comment by arily** [comment]
I think you can squash the commits. It's a little bit cluttered rn

**Comment by arily** [comment]
I think it's ready now

**Review by cmyui — APPROVED** [review]

**Comment by cmyui** [comment]
thanks for the pr!

**Pull request merged**